src/core/security/credentials.c

```diff
@@ -871,9 +871,12 @@ static void fake_oauth2_get_request_metadata(grpc_credentials *creds,
   grpc_fake_oauth2_credentials *c = (grpc_fake_oauth2_credentials *)creds;
 
   if (c->is_async) {
-    grpc_iomgr_add_callback(
-        on_simulated_token_fetch_done,
+    grpc_iomgr_closure *on_simulated_token_fetch_done_closure =
+        gpr_malloc(sizeof(grpc_iomgr_closure));
+    grpc_iomgr_managed_closure_init(
```

Comment:
This might take an iteration or two with @jboeuf - but I think the changes we were talking about yesterday evening might help here.

I'm worried that we're going to spam auth servers if we are under load with the current architecture. Imagine running 10kqps and having an auth server for token fetches with 10ms latency (which is not at all unbelievable) - now we're doing 100 identical requests against the auth server, and not making that latency any better.

I'd prefer to see something in either auth.c or credentials.c that guarantees that at most one request for new credential tokens is made at any given time. If we can get that, then this path gets a little cleaner in that we can preallocate the closure object as we do in the rest of the codebase.